index.src.html

```diff
@@ -749,6 +755,166 @@ <h3 id="try-delivery" algorithm>
   8.  Return "`Failure`".
 </section>
 
+<section>
+  <h2 id="observers">Reporting Observers</h2>
+
+  A <dfn>reporting observer</dfn> observes some types of <a>reports</a> from
+  JavaScript, and is represented in JavaScript by the {{ReportingObserver}}
+  object.
+
+  Each <a>environment settings object</a> has a <dfn>registered reporting
+  observer list</dfn>, which is an <a>ordered set</a> of <a>reporting observers</a>.
+
+  Any <a>reporting observer</a> that is in a <a>registered reporting observer
+  list</a> is considered <dfn>registered</dfn>.
+
+  Each <a>environment settings object</a> has a <dfn>report buffer</dfn>, which
+  is an ordered list of <a>reports</a> that have been generated in
+  that <a>environment settings object</a>. This list is initially empty, and the
+  reports are stored in the same order in which they are generated.
+
+  <h3 id=interface-reporting-observer>Interface {{ReportingObserver}}</h3>
+
+  <pre class="idl">
+[NoInterfaceObject]
+interface ReportBody {
+};
+
+[NoInterfaceObject]
+interface Report {
+  readonly attribute DOMString type;
+  readonly attribute DOMString url;
+  readonly attribute ReportBody? body;
+};
+
+[Constructor(ReportingObserverCallback callback, optional ReportingObserverOptions options)]
+interface ReportingObserver {
+  void observe();
+  void disconnect();
+};
+
+callback ReportingObserverCallback = void (sequence&lt;Report> reports, ReportingObserver observer);
+
+dictionary ReportingObserverOptions {
+  sequence&lt;DOMString> types;
+  boolean futureOnly = false;
+};
+  </pre>
+
+  Issue: How do we specify which reports are observable?  Network-layer reports
```

Comment:
It seems that each report type should be explicitly stating if it's meant to be JS-observable. We can make this a flag on the ~queue the report step, with default value set to "false".

Reply:
Okay, I've added a step to check for this. Each specific report type will need to explicitly state that it is observable from JavaScript in order to be observed by a ReportingObserver.

Reply:
The parameter would need to be per-report, not per-type.  Only **some** Network Error Logging reports should be visible to observers, for instance.  In particular, a NEL report should only be visible to the same origin that the report is about.  So for instance, if a page at `foo.com` contains a link to `bar.com`, `bar.com` can request NEL reports, but we **won't** send those reports about `bar.com` to any `foo.com` Reporting collectors.  And so any Reporting observers registered for `foo.com` shouldn't be able to see those `bar.com` NEL reports, either.

I **think** we can still get that with the parameter that the downstream spec has to set, indicating whether individual reports are visible to observers or not.  NEL should have all the info needed to fill that in — if the environment settings object's origin is different than the origin of the request that the report describes, set it to `no`; otherwise set it to `yes`.  I'd appreciate someone checking my work on that, though.  :wink: 

So in addition to the parameter being present, I'd suggest some clarifying text.  ("Downstream specs should make sure to set this parameter so that observers are only able to see reports that would be sent to collectors for that origin" or something like that)